### ISSUE TITLE
Produce usable bounding boxes in script detection again

### ISSUE THREAD

**Issue by mittagessen** [mittagessen/kraken#98]
pytorch's new CTC loss does not produce the drawn out activations in the softmax layer that Tom Breuel's implementation did. Currently the pytorch-trained script detection network cuts off part of the first and final grapheme of a single-script run.

A simple connected component-based expansion with some heuristics for conflicting labels might be enough here.

**Comment by babraham123** [comment]
Please fix

**Comment by tbaptista** [comment]
Is there any idea on when this issue might be fixed? If given some guidance, I could help.

**Comment by mittagessen** [comment]
It's really not high priority as it isn't crucial (or probably possible) for recognition of manuscripts. As I've written the problem is that the new activations are too peaked to be used as accurate bounding boxes and they have to be expanded somehow.

If you like you can take a look at the current implementation of `detect_scripts` in `pageseg.py`. A good start would probably be to just greedily expand all bounds to the midpoint of its neighbors (just iterate over the preds array at the end and recalculate the boxes) and if that doesn't work try something fancier like connected components to sort it out.

I've just started retraining the model and will upload it in a few hours if you'd like to experiment.

**Comment by mittagessen** [comment]
[Here](http://homer.dh.uni-leipzig.de/script_2304_best.mlmodel)'s the freshly trained model.

**Comment by tbaptista** [comment]
Thanks for the reply.

Actually, I misunderstood the issue and thought it was happening at the cuts of the rpred, not the detect_script. I am having a similar issue there with a custom trained model. Essentially the width of the bounding boxes is always much lower than the character width. Is it possible the two issues are related?

**Comment by mittagessen** [comment]
Yes, the issues stem from the same source as the script detection is just another recognition model. For recognition we (mostly) don't care as the character activations were always somewhat fuzzy and not usable for e.g. cutting out single characters from the page (this was less of an issue with the script detector net). Something solving either issue should work for the other, albeit doing it in a script independent manner might require some work.

**Comment by bertsky** [comment]
Suggestion: instead of keeping whitespace (treating it as explicit output element) and cutting words at the decoded intervals (start to stop), combine word start and follow-up whitespace start. For the last token, use the line stop.
